src/NelsonSiegelSvensson.jl

```diff
@@ -0,0 +1,179 @@
+
+    #function NelsonSiegelSvensson(β₀, β₁, β₂, β₃, τ₁, τ₂, r²)
+    function NelsonSiegelSvensson(β₀, β₁, β₂, β₃, τ₁, τ₂)
+        #(τ₁ <= 0 || τ₂ <= 0 || r² < 0 || r² > 1) && throw(DomainError("Wrong parameter ranges"))
+        (τ₁ <= 0 || τ₂ <= 0) && throw(DomainError("Wrong parameter ranges"))
+        #return new(β₀, β₁, β₂, β₃, τ₁, τ₂, r²)
+        return new(β₀, β₁, β₂, β₃, τ₁, τ₂)
+    end
+end
+
+Base.zero(ns::NelsonSiegel, t) = ns.β₀ .+ ns.β₁ .* (1.0 .- exp.(-t ./ ns.τ₁)) ./ (t ./ ns.τ₁) .+ ns.β₂ .* ((1.0 .- exp.(-t ./ ns.τ₁)) ./ (t ./ ns.τ₁) .- exp.(-t ./ ns.τ₁))
+discount(ns::NelsonSiegel, t) = exp.(-t .* zero(ns, t))
+
+Base.zero(nss::NelsonSiegelSvensson, t) = nss.β₀ .+ nss.β₁ .* (1.0 .- exp.(-t ./ nss.τ₁)) ./ (t ./ nss.τ₁) .+ nss.β₂ .* ((1.0 .- exp.(-t ./ nss.τ₁)) ./ (t ./ nss.τ₁) .- exp.(-t ./ nss.τ₁)) .+ nss.β₃ .* ((1.0 .- exp.(-t ./ nss.τ₂)) ./ (t ./ nss.τ₂) .- exp.(-t ./ nss.τ₂))
+discount(nss::NelsonSiegel, t) = exp.(-t .* zero(nss, t))
+
+using LsqFit
+
+""" 
+    est_ns_params(swq::Vector{SwapQuote}, τₐₗₗ::Array{Float, 1})
+
+Return the NelsonSiegel fitted parameters.
+"""
+function est_ns_params(swq::Vector{SwapQuote}, τₐₗₗ::Array{Float64, 1} = [0.1, 0.15, 0.2, 0.3, 0.5, 0.75, 1, 1.5, 2, 3, 5, 7.5, 10]) where {Q<:ObservableQuote}
+    # τₐₗₗ = [0.1, 0.15, 0.2, 0.3, 0.5, 0.75, 1, 1.5, 2, 3, 5, 7.5, 10] move to parameter to allow more freedom
+    yields = [q.yield / q.frequency for q in swq] # assume simple interests
+    maturities = [q.maturity for q in swq]
+    Δₘ = vcat([maturities[1]], diff(maturities))
+    #sp = sum(prices .* prices)
+    total_resid = Inf
+    ns_param = NelsonSiegel(1.0, 0.0, 0.0, 1)
+
+    for τ in τₐₗₗ
+        spot(m, param) = zero(NelsonSiegel(param[1], param[2], param[3], param[4]), m)
+        param₀ = [1.0, 0.0, 0.0]
+        res = LsqFit.curve_fit(spot, maturities, yields, Δₘ, param₀)
+        sr = sum(res.resid .* res.resid)
+        if sr < total_resid # take the smallest sum of squares of residuals
+            total_resid = sr
+            #r² = 1.0 - sr / sp
+            ns_param = NelsonSiegel(res.param[1], res.param[2], res.param[3], τ)
+        end
+    end
+
+    return ns_param
+end
+
+""" 
+    est_ns_params(yields::AbstractVector, maturities::AbstractVector, τₐₗₗ::Array{Float64, 1})
+
+Return the NelsonSiegel fitted parameters.
+"""
+function est_ns_params(yields::AbstractVector, maturities::AbstractVector, τₐₗₗ::Array{Float64, 1} = [0.1, 0.15, 0.2, 0.3, 0.5, 0.75, 1, 1.5, 2, 3, 5, 7.5, 10])
+    # τₐₗₗ = [0.1, 0.15, 0.2, 0.3, 0.5, 0.75, 1, 1.5, 2, 3, 5, 7.5, 10] move to parameter to allow more freedom
+    Δₘ = vcat([maturities[1]], diff(maturities))
+    #sp = sum(prices .* prices)
+    total_resid = Inf
+    ns_param = NelsonSiegel(1.0, 0.0, 0.0, 1)
+
+    for τ in τₐₗₗ
+        spot(m, param) = zero(NelsonSiegel(param[1], param[2], param[3], param[4]), m)
+        param₀ = [1.0, 0.0, 0.0]
+        res = LsqFit.curve_fit(spot, maturities, yields, Δₘ, param₀)
```

Comment:
Again, I don't have experience with `LsqFit`, but I think it does some advanced stuff. In this case, however, you're doing a linear least squares, so maybe there's a simpler way. I'm not sure, but give it a thought.

Reply:
Agree with the thoughts above, but not a big deal to add a somewhat light dependency like this. It looks like everywhere `LsqFit` is used it's qualified with `LsqFit.functionname`. That means instead of `using LsqFit` it could be `import LsqFit` which is probably better code hygiene we should try to use since it doesn't pollute the namespace and it's really clear which functions are defined in other packages.